maven-core/src/main/java/org/apache/maven/graph/MultiModuleCollectionStrategy.java

```diff
@@ -0,0 +1,181 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.execution.MavenExecutionRequest;
+import org.apache.maven.model.Plugin;
+import org.apache.maven.model.building.ModelProblem;
+import org.apache.maven.plugin.PluginManagerException;
+import org.apache.maven.plugin.PluginResolutionException;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectBuildingException;
+import org.apache.maven.project.ProjectBuildingResult;
+import org.codehaus.plexus.logging.Logger;
+import org.eclipse.aether.resolution.ArtifactResolutionException;
+import org.eclipse.aether.transfer.ArtifactNotFoundException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+/**
+ * Strategy for collecting Maven projects from the multi-module project root, even when executed in a submodule.
+ */
+@Named( "MultiModuleCollectionStrategy" )
+@Singleton
+public class MultiModuleCollectionStrategy implements ProjectCollectionStrategy
+{
+    private final Logger logger;
+    private final ProjectCollector projectCollector;
+
+    @Inject
+    public MultiModuleCollectionStrategy( Logger logger, ProjectCollector projectCollector )
+    {
+        this.logger = logger;
+        this.projectCollector = projectCollector;
+    }
+
+    @Override
+    public List<MavenProject> collectProjects( MavenExecutionRequest request ) throws ProjectBuildingException
+    {
+        File moduleProjectPomFile = getMultiModuleProjectPomFile( request );
+        List<File> files = Collections.singletonList( moduleProjectPomFile.getAbsoluteFile() );
+        try
+        {
+            List<MavenProject> projects = new ArrayList<>();
+            projectCollector.collectProjects( projects, files, request );
+            boolean isRequestedProjectCollected = isRequestedProjectCollected( request, projects );
+            if ( isRequestedProjectCollected )
+            {
+                return projects;
+            }
+            else
+            {
+                return Collections.emptyList();
+            }
+        }
+        catch ( ProjectBuildingException e )
+        {
+            boolean fallThrough = isModuleOutsideRequestScopeDependingOnPluginModule( request, e );
+
+            if ( fallThrough )
+            {
+                return Collections.emptyList();
+            }
+
+            throw e;
+        }
+    }
+
+    private File getMultiModuleProjectPomFile( MavenExecutionRequest request )
+    {
+        if ( request.getPom().getParentFile().equals( request.getMultiModuleProjectDirectory() ) )
+        {
+            return request.getPom();
+        }
+        else
+        {
+            File multiModuleProjectPom = new File( request.getMultiModuleProjectDirectory(), "pom.xml" );
```

Comment:
This might not work with [polyglot-maven](https://github.com/takari/polyglot-maven). IIRC there's a PomLocator that handles the right extension.

Reply:
I found a `DefaultModelLocator` but its default implementation does exactly this:

```java
return new File( projectDirectory, "pom.xml" );
```

I found `PomFinder` but that's definitely doing something else.

So maybe there's another class for this? Any hints for its name?

Reply:
So that means that if we have an AtomModelLocator, it'll select a different pom, so I guess that's the right class. Keep in mind, that if there's a non-default ModelLocator, that's the one that will be picked, otherwise Maven will fall back to this default.

Reply:
Just to make sure I understand correctly: the `ModelLocator` interface is the one we need here?

Reply:
Yes, you will need the ModelLocator Component, the right implementation will be serverd for you.